agr_java_core/src/main/java/org/alliancegenome/core/translators/document/PhenotypeTranslator.java

```diff
@@ -0,0 +1,281 @@
+                            .filter(entry -> phenotype != null)
+                            .collect(Collectors.groupingBy(PhenotypeEntityJoin::getFeature
+                            ));
+                    if (!featurelessJoins.isEmpty())
+                        featureMap.put(null, featurelessJoins);
+                    return featureMap.entrySet().stream()
+                            .map(featureMapEntry -> {
+                                PhenotypeAnnotationDocument document = new PhenotypeAnnotationDocument();
+                                document.setPhenotype(phenotype.getPhenotypeStatement());
+                                document.setGeneDocument(geneTranslator.translate(geneMapEntry.getKey(), 0));
+                                Feature feature = featureMapEntry.getKey();
+                                if (feature != null) {
+                                    document.setFeatureDocument(featureTranslator.translate(feature, 0));
+                                }
+                                document.setPublications(getPublicationDoclets(featureMapEntry.getValue()));
+                                return document;
+                            }).collect(Collectors.toList());
+                })
+                .flatMap(Collection::stream)
+                .collect(Collectors.toList());
+    }
+
+    private Map<Gene, List<PhenotypeEntityJoin>> getGeneAnnotationMap(Phenotype phenotype, Gene gene) {
+        // group by gene then by association type
+        Map<Gene, List<PhenotypeEntityJoin>> geneMap = phenotype.getPhenotypeEntityJoins().stream()
+                .filter(phenotypeEntityJoin -> gene == null || phenotypeEntityJoin.getGene().equals(gene))
+                .collect(groupingBy(PhenotypeEntityJoin::getGene));
+
+        // sort by gene symbol
+        return geneMap.entrySet().stream()
+                .sorted(Map.Entry.comparingByKey())
+                .collect(toMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    public Map<Gene, Map<Optional<Feature>, List<PhenotypeEntityJoin>>> getGeneFeatureAnnotationMap(Phenotype phenotype, Gene gene) {
+        // group by gene then by feature
+        Map<Gene, Map<Optional<Feature>, List<PhenotypeEntityJoin>>> geneMap = phenotype.getPhenotypeEntityJoins().stream()
+                .filter(phenotypeEntityJoin -> gene == null || phenotypeEntityJoin.getGene().equals(gene))
+                .collect(
+                        groupingBy(PhenotypeEntityJoin::getGene,
+                                groupingBy(join -> Optional.ofNullable(join.getFeature())))
+                );
+
+        // sort by gene symbol
+        return geneMap.entrySet().stream()
+                .sorted(Map.Entry.comparingByKey())
+                .collect(toMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    private PublicationDoclet getPublicationDoclet(PhenotypeEntityJoin association, Publication publication) {
```

Comment:
Can this be turned into a Doclet Translator this code also exists in the DiseaseTranslator.java?

Reply:
In the DiseaseTranslator it uses DiseaseEntityJoin instead of PhenotypeEntityJoin... It also uses the DOTerm entity instead of Phenotype.
I could create a super class called EntityJoin that contains gene and feature and publication and would leave the child class just adding the main entity - DoTerm or Phenotype...

Reply:
In this method, only the publication is used. I was suggesting that Publication needs to be a translator. Plus why is association passed in and not used? Maybe the evidences codes could be passed in as a list?